src/mlpack/methods/ann/layer/repeat_impl.hpp

```diff
@@ -0,0 +1,204 @@
+/**
+ * @file methods/ann/layer/repeat_impl.hpp
+ * @author Adam Kropp
+ *
+ * Implementation of the Repeat class, which repeats the input n times
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LAYER_REPEAT_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_REPEAT_IMPL_HPP
+
+#include "repeat.hpp"
+
+namespace mlpack {
+
+template<typename MatType>
+RepeatType<MatType>::RepeatType(
+    std::vector<size_t> multiples, bool interleave) :
+    Layer<MatType>(),
+    multiples(std::move(multiples)),
+    interleave(interleave)
+{
+  // Nothing to do.
+}
+
+template<typename MatType>
+RepeatType<MatType>::RepeatType() :
+    Layer<MatType>(),
+    interleave(false)
+{
+  // Nothing to do.
+}
+
+template<typename MatType>
+RepeatType<MatType>::RepeatType(const RepeatType& other) :
+    Layer<MatType>(other),
+    multiples(other.multiples),
+    interleave(other.interleave),
+    outIdxs(other.outIdxs),
+    sizeMult(other.sizeMult),
+    backIdxs(other.backIdxs)
+{
+  // Nothing else to do.
+}
+
+template<typename MatType>
+RepeatType<MatType>::RepeatType(RepeatType&& other)  noexcept :
+    Layer<MatType>(std::move(other)),
+    multiples(other.multiples),
+    interleave(other.interleave),
+    outIdxs(other.outIdxs),
+    sizeMult(other.sizeMult),
+    backIdxs(other.backIdxs)
+{
+  // Nothing else to do.
+}
+
+template<typename MatType>
+RepeatType<MatType>& RepeatType<MatType>::operator=(const RepeatType& other)
+{
+  if (this != &other)
+  {
+    Layer<MatType>::operator=(other);
+    multiples = other.multiples;
+    interleave = other.interleave;
+    outIdxs = other.outIdxs;
+    sizeMult = other.sizeMult;
+    backIdxs = other.backIdxs;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+RepeatType<MatType>& RepeatType<MatType>::operator=(RepeatType&& other) noexcept
+{
+  if (this != &other)
+  {
+    Layer<MatType>::operator=(std::move(other));
+    multiples = std::move(other.multiples);
+    interleave = std::move(other.interleave);
+    outIdxs = std::move(other.outIdxs);
+    sizeMult = other.sizeMult;
+    backIdxs = other.backIdxs;
+  }
+
+  return *this;
+}
+
+template<typename MatType>
+void RepeatType<MatType>::ComputeOutputDimensions()
+{
+  if (multiples.size() > this->inputDimensions.size())
+  {
+    std::ostringstream oss;
+    oss << "Repeat::ComputeOutputDimensions(): multiples vector must "
+        << "have the same or fewer dimensions than InputDimensions";
+    throw std::invalid_argument(oss.str());
+  }
+
+  size_t inputSize = this->inputDimensions[0];
+  for (size_t i = 1; i < this->inputDimensions.size(); i++)
+  {
+    inputSize *= this->inputDimensions[i];
+  }
+  arma::umat idxs = arma::regspace<arma::uvec>(0, inputSize - 1);
```

Comment:
I think this needs to follow the `MatType` @rcurtin correct me If I am wrong. If I understand correctly this matrix here is to define the dimension of the output. Suppose we are using this with Bandicoot, then we need to create the matrix on the GPU.

I would change this line to make it generic `MatType` in addition to use `ColVecType` inside instead of `arma::uvec`

Reply:
@akropp If this should not affect the weights in the layer, then it could pass. However, I would prefer to avoid this, @rcurtin your call here

Reply:
Bandicoot doesn't have `regspace`; I know we will need to do adaptation later, but personally I don't think that we need to do it now.  It's nice when we can easily use ADL and strip the `arma::`, but at least for this one I think it's too much work and not worth it yet---we can come back later once we have the necessary Bandicoot support.